tests/check_eventloop_udp.c

```diff
@@ -0,0 +1,385 @@
+        UA_Variant_setScalar(&params[1].value, &targetHost, &UA_TYPES[UA_TYPES_STRING]);
+
+        UA_ConnectionManager *cmWriter = UA_ConnectionManager_new_POSIX_UDP(UA_STRING("udpCM"));
+
+        /* the cm writer should not need a connection callback */
+        cmWriter->connectionCallback = connectionCallback;
+        elWrite->registerEventSource(elWrite, &cmWriter->eventSource);
+        elWrite->start(elWrite);
+        UA_StatusCode retval = cmWriter->openConnection(cmWriter, 2, params, (void*)0x01);
+
+        ck_assert_uint_eq(retval, UA_STATUSCODE_GOOD);
+        for(size_t i = 0; i < 2; i++) {
+            UA_DateTime next = el->run(el, 1);
+            UA_fakeSleep((UA_UInt32)((next - UA_DateTime_now()) / UA_DATETIME_MSEC));
+        }
+        /* The reader el should not generate any connections */
+        ck_assert_uint_eq(connCount, 0);
+
+        /* The writer el should receive a signal "ready to be written on" */
+        ck_assert_uint_eq(retval, UA_STATUSCODE_GOOD);
+        for(size_t i = 0; i < 2; i++) {
+            UA_DateTime next = elWrite->run(elWrite, 1);
+            UA_fakeSleep((UA_UInt32)((next - UA_DateTime_now()) / UA_DATETIME_MSEC));
+        }
+        ck_assert_uint_eq(connCount, 1);
+
+        /* Send a message from the writer */
+        received = false;
+        UA_ByteString snd;
+        retval = cmWriter->allocNetworkBuffer(cmWriter, clientId, &snd, strlen(testMsg));
+        ck_assert_uint_eq(retval, UA_STATUSCODE_GOOD);
+        memcpy(snd.data, testMsg, strlen(testMsg));
+        retval = cmWriter->sendWithConnection(cmWriter, clientId, 0, NULL, &snd);
+        ck_assert_uint_eq(retval, UA_STATUSCODE_GOOD);
+        for(size_t i = 0; i < 2; i++) {
+            UA_DateTime next = el->run(el, 1);
+            UA_fakeSleep((UA_UInt32)((next - UA_DateTime_now()) / UA_DATETIME_MSEC));
+        }
+        ck_assert(received);
+
+        /* Close the connection */
+        retval = cmWriter->closeConnection(cmWriter, clientId);
+        ck_assert_uint_eq(retval, UA_STATUSCODE_GOOD);
+        ck_assert_uint_eq(connCount, 1);
+        for(size_t i = 0; i < 2; i++) {
+            UA_DateTime next = elWrite->run(elWrite, 1);
+            UA_fakeSleep((UA_UInt32)((next - UA_DateTime_now()) / UA_DATETIME_MSEC));
+        }
+        ck_assert_uint_eq(connCount, 0);
+
+        /* Stop the Write EventLoop */
+        int max_stop_iteration_count = 10;
+        int iteration = 0;
+        /* Stop the EventLoop */
+        elWrite->stop(elWrite);
+        while(elWrite->state != UA_EVENTLOOPSTATE_STOPPED &&
+              iteration < max_stop_iteration_count) {
+            UA_DateTime next = elWrite->run(elWrite, 1);
+            UA_fakeSleep((UA_UInt32)((next - UA_DateTime_now()) / UA_DATETIME_MSEC));
+            iteration++;
+        }
+        ck_assert(elWrite->state == UA_EVENTLOOPSTATE_STOPPED);
+        elWrite->free(elWrite);
+        elWrite = NULL;
+
+        /* Stop the Read EventLoop */
```

Comment:
Add an assert that the message was received by the reader and the message content matches.
For example by setting a global flag in the receive callbacks.

Reply:
The assert does already happen in the `connectionCallback` of the `listener` and the assert does already happen after the listener eventloop did its iterations following the `sendWithConnection` via `ck_assert(received)`.

Or do you mean something else?

Reply:
Got it, all good.
I am fine with the PR.